core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -650,17 +651,55 @@ private void applyInitializingShard(final ClusterState state, final IndexMetaDat
             final RecoveryState recoveryState = new RecoveryState(indexShard.shardId(), shardRouting.primary(),
                 RecoveryState.Type.STORE,
                 nodes.getLocalNode(), nodes.getLocalNode());
-            indexShard.markAsRecovering("from store", recoveryState); // mark the shard as recovering on the cluster state thread
-            threadPool.generic().execute(() -> {
-                try {
-                    if (indexShard.recoverFromStore(nodes.getLocalNode())) {
-                        shardStateAction.shardStarted(shardRouting, "after recovery from store", SHARD_STATE_ACTION_LISTENER);
+            Index mergeSourceIndex = indexShard.indexSettings().getIndexMetaData().getMergeSourceIndex();
+            if (mergeSourceIndex != null) {
+                IndexService sourceIndexService = indicesService.indexService(mergeSourceIndex);
+                int numShards = sourceIndexService.getMetaData().getNumberOfShards();
+                List<IndexShard> startedShards = new ArrayList<>();
+                for (IndexShard shard : sourceIndexService) {
+                    if (shard.state() == IndexShardState.STARTED) {
+                       startedShards.add(shard);
+                    } else if (shard.state() == IndexShardState.POST_RECOVERY) {
```

Comment:
do we really need to allow for this and have the added complexity? did you have to do this because of the order in which apply the *same* cluster state , causing the source shards to be moved to started after the initializing shard? we can maybe change the order of applying the cluster state, or otherwise just assume it's ok here (maybe add an assertion) and let thing explode later on when the shards tell us their engines are not started?